Fix a crash during SMT translation of structs

SaswatPadhi opened the pull request:
This PR fixes a bug with SMT translation of structs using data types (not bit vectors). CBMC `--smt2 --z3` was crashing on the attached regression test with a `key not found` exception from a `map::at` call.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- NA ~~Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.~~
- NA ~~The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/~~
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- NA ~~My commit message includes data points confirming performance improvements (if claimed).~~
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

SaswatPadhi committed:
Fix SMT2 DT encoding of structs

This commit fixes a `key not found` exception in `convert_type` during SMT2 encoding of structs using data types. Although structs were being mapped to SMT2 datatypes (in `datatype_map`) struct_tags were not.

martin-cs commented on regression/cbmc-smt2-z3/Makefile:
```diff
@@ -0,0 +1,19 @@
+include ../../src/common
+
+test: ../test.pl
+	@../test.pl -e -p -c "../../../src/cbmc/cbmc --smt2 --z3"
```
Rather than adding a separate directory, can you tag the test.desc as `smt-backend` as is done in `cbmc/integer-assignments1/test.desc`

SaswatPadhi replied:
Hi @martin-cs,

I just tagged it with `smt-backend` and ran locally, but I see the `smt-backend` tests are run with `--cprover-smt2` not with `--smt2`.

This regression test was crashing earlier with `cbmc --smt2` and now it runs as expected with that flag. However, the test does not pass (gives `VERIFICATION ERROR`) with `cbmc --cprover-smt2` -- same without the changes introduced by this PR.

So, I am not sure what's the best way to tag this test. I have kept it as `CORE` but have added the appropriate flags in `test.desc`. What do you suggest?

martin-cs commented on regression/cbmc-smt2-z3/struct/test.desc:
```diff
@@ -0,0 +1,9 @@
+
+^EXIT=6$
+^SIGNAL=0$
+^VERIFICATION ERROR$
```
If this fixes the translation, why is the test case giving an error?

SaswatPadhi replied:
Oh this is because the assertion doesn't hold.

With a trivial property, CBMC was just simplifying it to `true` and was not even generating SMT queries. I can try to check a different property that won't be trivially simplified to true and would actually require an SMT solver. Or, can I disable all optimizations using some flags?

tautschnig replied:
If the assertion doesn't hold, then the result should be `VERIFICATION FAILED`, not `VERIFICATION ERROR`?!

To disable constant propagation and/or simplification: use `--no-propagation` and `--no-simplify`, respectively.

kroening replied:
I would avoid either one of these options unless they are specifically tested.
If it bugs you that this is a satisfiable instance (which it shouldn't) then you could use s.th. like
`if(t.a>10) assert(t.a > 0);`

SaswatPadhi replied:
Thanks everyone! I implemented Daniel's suggestion above. Marking this one as resolved.

martin-cs reviewed (changes requested):
Please can you reformat this a little to fix with current conventions.

kroening commented:
Should there be a similar change for union_tags?

SaswatPadhi commented:
> Should there be a similar change for union_tags?

Unions are currently not encoded using data types in SMT, so this particular issue doesn't appear with unions.

But I haven't checked if we would have a similar issue with non-DT encoding of unions and structs. We do something similar for bit vector encodings -- instead of looking up the `datatype_map`, we look up `boolbv_width` of the type. I will check this and create a separate PR to fix the BV encoding, if needed.

SaswatPadhi commented:
I think the test is failing in CI because it cannot find `z3`.

Is there a way we can just test the SMT constraint generation with data types enabled, without actually solving them using z3? Alternatively, we can install z3 on our CI :)

thomasspriggs commented:
> I think the test is failing in CI because it cannot find `z3`.
> 
> Is there a way we can just test the SMT constraint generation with data types enabled, without actually solving them using z3? Alternatively, we can install z3 on our CI :)

It seems reasonable to install an SMT solver in CI in order to test the SMT output properly. Just so long as we don't end up giving it anything to solve which inflates the overall run time too much. @diffblue/diffblue-opensource What do the rest of you think?

Please note that I haven't reviewed the rest of this PR, I am just trying to help with this particular issue.

SaswatPadhi commented:
Thanks @thomasspriggs.

@martin-cs I think I have addressed your comment regarding conventions. There is one remaining issue with `smt-backend` tag which I have described above. Could you please take another look?

thomasspriggs commented on src/solvers/smt2/smt2_conv.cpp:
```diff
@@ -4869,8 +4869,10 @@ void smt2_convt::find_symbols_rec(
-      find_symbols_rec(ns.follow_tag(struct_tag), recstack);
+      find_symbols_rec(base_struct, recstack);
+      datatype_map[type] = datatype_map[base_struct];
     }
```
:+1: I have tested this locally and this does indeed seem to fix a failure wherein cbmc prints `map::at` and exits.

SaswatPadhi replied:
Thanks.

thomasspriggs commented on src/solvers/smt2/smt2_conv.cpp:
```diff
@@ -4869,8 +4869,10 @@ void smt2_convt::find_symbols_rec(
       recstack.insert(id);
-      find_symbols_rec(ns.follow_tag(struct_tag), recstack);
+      find_symbols_rec(base_struct, recstack);
+      datatype_map[type] = datatype_map[base_struct];
```
:pick: It would be nice to have some more doxygen at the declaration of `datatype_map` covering exactly what it maps to/from. It would also be nice to have better doxygen above `use_datatypes` documenting what changes should be expected in the output file when this is set to true. Obviously not a blocking comment, just wishful thinking.

SaswatPadhi replied:
I'm adding some comments at `datatype_map` declaration, but I think I could make a separate PR for explaining `use_datatypes` better.

There's some documentation about `use_datatypes` sprinkled throughout `smt2_conv.cpp`, e.g. at https://github.com/diffblue/cbmc/blob/develop/src/solvers/smt2/smt2_conv.cpp#L4775-L4810. Probably we could add some high-level description above its declaration.

thomasspriggs commented on regression/cbmc/struct15/test.desc:
```diff
@@ -0,0 +1,9 @@
+^EXIT=0$
+^SIGNAL=0$
+^VERIFICATION SUCCESSFUL$
+--
```
:pick: Might be nice to have a regex in the must not be matched section here checking for `map::at`.

SaswatPadhi replied:
Good idea, thanks! Pushed this change.

thomasspriggs reviewed (commented):
The fix appears reasonable to me. But we need get the test to run and pass in CI, before I will be happy to approve. This so that the additional CI related changes can be reviewed and approved before merging.

SaswatPadhi commented:
Blocked on #5949.

thomasspriggs commented:
> Blocked on #5949.

I only raised that PR to run CI, it is in Draft status and the work in known to be incomplete. We are currently giving higher priority to other pieces of work. It could be a week or so, before we come back to those CI changes. Is that going to be a problem for you?

SaswatPadhi commented:
Hi @thomasspriggs,

No problem. I was going through all of my PRs and rebased some of them with `develop`. This is just a note to myself. I have things running locally for demos etc. so this is not a high priority for now.

Thanks.

martin-cs reviewed (approved):
Looks good; thanks.

peterschrammel commented on regression/cbmc/struct15/test.desc:
```diff
@@ -0,0 +1,11 @@
+CORE
+main.c
+--trace --smt2
```
I don't think this code is even exercised without `--z3`, see https://github.com/diffblue/cbmc/blob/develop/src/solvers/smt2/smt2_conv.cpp#L101

thomasspriggs replied:
If `--smt2` is specified and a specific solver is not specified, then z3 is selected by default. See https://github.com/diffblue/cbmc/blob/develop/src/cbmc/cbmc_parse_options.cpp#L428 and https://github.com/diffblue/cbmc/blob/develop/src/cbmc/cbmc_parse_options.cpp#L438
That said, I think this would be a more robust test if `--z3` was specified, rather than relying on the default.

SaswatPadhi replied:
Thanks, that's a good point! I will make this change.
